File: LEETCODE/Graphs/traversalProblems/S2_ConnectedComponents.java
package LEETCODE.Graphs.traversalProblems;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class S2_ConnectedComponents {

    // BFS to visit all nodes in one connected component
    private static void bfs(int node, boolean vis[], ArrayList<ArrayList<Integer>> adj) {
        Queue<Integer> q = new LinkedList<>();
        // Start BFS from the given node
        q.offer(node);
        vis[node] = true;
        while (!q.isEmpty()) {
            int nde = q.poll();
            // Traverse all neighbours of current node
            for (int nei : adj.get(nde)) {
                if (!vis[nei]) {
                    vis[nei] = true;
                    q.offer(nei);
                }
            }
        }
    }
    public static int findNumberOfComponent(int V, List<List<Integer>> edges) {
        // ---------------------------------------------------
        // Step 1 : Create Adjacency List
        // ---------------------------------------------------
        ArrayList<ArrayList<Integer>> adj = new ArrayList<>();
        for (int i = 0; i < V; i++) {
            adj.add(new ArrayList<>());
        }
        // Since graph is undirected, add both directions
        for (int i = 0; i < edges.size(); i++) {
            int u = edges.get(i).get(0);
            int v = edges.get(i).get(1);
            adj.get(u).add(v);
            adj.get(v).add(u);
        }
        // ---------------------------------------------------
        // Step 2 : BFS on every unvisited node
        // Every new BFS = One Connected Component
        // ---------------------------------------------------
        boolean vis[] = new boolean[V];
        int count = 0;
        for (int i = 0; i < V; i++) {
            // New component found
            if (!vis[i]) {
                count++;
                // Visit the complete component
                bfs(i, vis, adj);
            }
        }
        return count;
    }
    public static void main(String[] args) {
        // Example:
        //
        // 0 ----- 1 ----- 2
        //
        // 3 ----- 4
        //
        // Connected Components:
        // {0,1,2}
        // {3,4}
        //
        // Answer = 2
        int V = 5;
        List<List<Integer>> edges = new ArrayList<>();
        edges.add(List.of(0, 1));
        edges.add(List.of(1, 2));
        edges.add(List.of(3, 4));
        int ans = findNumberOfComponent(V, edges);
        System.out.println("Number of Connected Components = " + ans);
    }
}

File: LEETCODE/Graphs/traversalProblems/S2_ConnectedComponents.cpp
#include<bits/stdc++.h>
using namespace std;

void bfs(int node, vector<vector<int>>& adj, vector<int>& vis) {
    queue<int> q;
    q.push(node);
    vis[node] = 1;
    while(!q.empty()) {
        int curr = q.front();
        q.pop();
        for(auto nei : adj[curr]) {
            if(!vis[nei]) {
                vis[nei] = 1;
                q.push(nei);
            }
        }
    }
}

int connectedComponents(vector<vector<int>>& edges, int v, int m){
    vector<vector<int>> adj(v+1);
    for(int i = 0; i < m; i++){
        int u = edges[i][0];
        int v = edges[i][1];
        adj[u].push_back(v);
        adj[v].push_back(u);
    }
    vector<int> vis(v,0); // its will be 0 based nodes
    int count = 0;
    for(int i = 0;  i < v; i++){
        if(vis[i] == 0){
            count++;
            bfs(i,adj,vis);
        }
    }
}
int main(){
    int n;
    int m;

    cin >> n >> m;
    vector<vector<int>> edges(m,vector<int>(2));
    for(int i = 0; i < m; i++){
        for(int j = 0; i < 2; j++){
            cin >>  edges[i][j];
        }
    }
}